## **Parsing**

In [2]:
from langchain_groq import ChatGroq

groq_model=ChatGroq(model="llama-3.3-70b-versatile")
groq_model.invoke("what is capital of USA?").content

'The capital of the United States of America (USA) is **Washington, D.C.** (short for District of Columbia).'

In [3]:
text="""In November 2023, xAI began previewing Grok as a chatbot to selected people,[10] with participation in the early access program being limited to paid X Premium users.[11]

It was announced that once the bot was out of early beta, it would only be available to higher tier X Premium+ subscribers.[12]

At the time of the preview, xAI described the chatbot as "a very early beta product – the best we could do with 2 months of training" that could "improve rapidly with each passing week".[13]

On March 11, 2024, Musk posted on X that the language model would go open source within a week. Six days later, on March 17, Grok-1 was open sourced under the Apache-2.0 license.[14][15] Disclosed were the networks architecture and its weight parameters.[16]

On March 26, 2024, Musk announced that Grok would be enabled for premium subscribers, not just those on the higher-end tier, Premium+.[17]

Grok-1.5
Grok-1.5
Developer(s)	xAI
Initial release	May 15, 2024; 10 months ago
Predecessor	Grok-1.5
Successor	Grok-2
Type	
Large language model
Foundation model
License	Proprietary
Website	x.ai/blog/grok-1.5
On March 29, 2024, Grok-1.5 was announced, with "improved reasoning capabilities" and a context length of 128,000 tokens.[18] Grok-1.5 was released to all X Premium users on May 15, 2024.[1]

On April 4, 2024, an update to X's "Explore" page included summaries of breaking news stories written by Grok, a task previously assigned to a human curation team.[19]

On April 12, 2024, Grok-1.5 Vision (Grok-1.5V) was announced. Grok-1.5V is able to process a wide variety of visual information, including documents, diagrams, graphs, screenshots, and photographs.[20] Grok-1.5V was never released to the public.

On May 4, 2024, Grok became available in the United Kingdom,[21] that being the only country in Europe to support Grok at the moment due to the impending Artificial Intelligence Act rules in the European Union. Grok was later reviewed by the EU and was released on May 16, 2024.[22]"""

In [4]:
from langchain_core.prompts import PromptTemplate

template=PromptTemplate(
    template="generate a prices 3 point summary from given text /n {text}",
    input_variables=["text"]
)

**Simple Chain**

In [8]:
chain = template | groq_model
chain=chain.invoke({"text":text})
print(chain)

content='Here is a 3-point summary of the prices of Grok:\n\n1. **Initial Access**: Initially, access to Grok was limited to paid X Premium users who participated in the early access program.\n2. **Tiered Access**: It was announced that once Grok was out of early beta, it would only be available to higher-tier X Premium+ subscribers, but this decision was later reversed, making it available to all X Premium users.\n3. **Current Availability**: Grok is currently available to all X Premium users, with no mention of additional costs or tiered pricing, although the exact pricing of X Premium is not specified in the provided text.' additional_kwargs={} response_metadata={'token_usage': {'completion_tokens': 134, 'prompt_tokens': 590, 'total_tokens': 724, 'completion_time': 0.795733098, 'prompt_time': 0.049924395, 'queue_time': 0.054279865000000004, 'total_time': 0.845657493}, 'model_name': 'llama-3.3-70b-versatile', 'system_fingerprint': 'fp_6507bcfb6f', 'finish_reason': 'stop', 'logprobs':

**Parsed Chain**

In [9]:
from langchain_core.output_parsers import StrOutputParser
parser = StrOutputParser()

chain_parser = template | groq_model | parser
chain_parser=chain_parser.invoke({"text":text})
print(chain_parser)

Here is a 3-point summary of the prices related to Grok:

1. **Initial Access**: Initially, access to Grok was limited to paid X Premium users who participated in the early access program.
2. **Tiered Access**: It was announced that once Grok was out of early beta, it would only be available to higher-tier X Premium+ subscribers, but this decision was later reversed, making it available to all X Premium users.
3. **Current Availability**: As of May 15, 2024, Grok-1.5 is available to all X Premium users, with no specific mention of additional costs or tiered pricing for access to the chatbot.


Check out the Change in response formate

---

**👉 Response from Simple Chain**

content='Here is a 3-point summary of the prices of Grok:\n\n1. **Initial Access**: Initially, access to Grok was limited to paid X Premium users who participated in the early access program.\n2. **Tiered Access**: It was announced that once Grok was out of early beta, it would only be available to higher-tier X Premium+ subscribers, but this decision was later reversed, making it available to all X Premium users.\n3. **Current Availability**: Grok is currently available to all X Premium users, with no mention of additional costs or tiered pricing, although the exact pricing of X Premium is not specified in the provided text.' additional_kwargs={} response_metadata={'token_usage': {'completion_tokens': 134, 'prompt_tokens': 590, 'total_tokens': 724, 'completion_time': 0.795733098, 'prompt_time': 0.049924395, 'queue_time': 0.054279865000000004, 'total_time': 0.845657493}, 'model_name': 'llama-3.3-70b-versatile', 'system_fingerprint': 'fp_6507bcfb6f', 'finish_reason': 'stop', 'logprobs': None} id='run--45a9d39e-2076-443f-9a62-1158bb05931f-0' usage_metadata={'input_tokens': 590, 'output_tokens': 134, 'total_tokens': 724}

---

**👉 Response from Parsed Chain**

Here is a 3-point summary of the prices related to Grok:

1. **Initial Access**: Initially, access to Grok was limited to paid X Premium users who participated in the early access program.
2. **Tiered Access**: It was announced that once Grok was out of early beta, it would only be available to higher-tier X Premium+ subscribers, but this decision was later reversed, making it available to all X Premium users.
3. **Current Availability**: As of May 15, 2024, Grok-1.5 is available to all X Premium users, with no specific mention of additional costs or tiered pricing for access to the chatbot.

## **JSON output Parser**

In [10]:
from langchain_core.output_parsers import JsonOutputParser

parser=JsonOutputParser()

In [11]:
parser.get_format_instructions()

'Return a JSON object.'

#### **Example 1**

In [48]:
template=PromptTemplate(
    template="give me name, age and city from the provided text {text} \n {format_instructions}" ,
    input_variables=['text'],
    partial_variables={"format_instructions": parser.get_format_instructions()}
)

text="""hi my name is Utkarsh Sinha my age is 29 and i am belong to bengaluru"""

**1️⃣⏩ Method One**

In [49]:
prompt=template.format(text=text)
prompt

result=groq_model.invoke(prompt)
print(result.content)

Here's a JSON object with the requested information:

```json
{
  "name": "Utkarsh Sinha",
  "age": 29,
  "city": "Bengaluru"
}
```


In [50]:
result=parser.parse(result.content)
print(result)

{'name': 'Utkarsh Sinha', 'age': 29, 'city': 'Bengaluru'}


In [51]:
name=result["name"]
age=result["age"]
city=result["city"]

print(name)
print(age)
print(city)

Utkarsh Sinha
29
Bengaluru


**2️⃣⏩ Method Two**

In [52]:
chain= template | groq_model | parser
chain_result=chain.invoke({"text": text})
print(chain_result)

{'name': 'Utkarsh Sinha', 'age': 29, 'city': 'Bengaluru'}


In [53]:
name=chain_result["name"]
age=chain_result["age"]
city=chain_result["city"]

print(name)
print(age)
print(city)

Utkarsh Sinha
29
Bengaluru


#### **Example Two**

In [55]:
topic = """AI agent or simply agent), expands this concept by proactively pursuing goals, making decisions, and taking actions over extended periods, thereby exemplifying a novel form of digital agency.[1]

Intelligent agents can range from simple to highly complex. A basic thermostat or control system is considered an intelligent agent, as is a human being, or any other system that meets the same criteria—such as a firm, a state, or a biome.[2]

Intelligent agents operate based on an objective function, which encapsulates their goals. They are designed to create and execute plans that maximize the expected value of this function upon completion.[3] For example, a reinforcement learning agent has a reward function, which allows programmers to shape its desired behavior.[4] Similarly, an evolutionary algorithm's behavior is guided by a fitness function.[5]

Intelligent agents in artificial intelligence are closely related to agents in economics, and versions of the intelligent agent paradigm are studied in cognitive science, ethics, and the philosophy of practical reason, as well as in many interdisciplinary socio-cognitive modeling and computer social simulations.

Intelligent agents are often described schematically as abstract functional systems similar to computer programs. To distinguish theoretical models from real-world implementations, abstract descriptions of intelligent agents are called abstract intelligent agents. Intelligent agents are also closely related to software agents—autonomous computer programs that carry out tasks on behalf of users. They are also referred to using a term borrowed from economics: a "rational agent".[2]"""

In [56]:
template2=PromptTemplate(
    template='Give me 5 facts about {topic} \n {format_instruction}',
    input_variables=['topic'],
    partial_variables={'format_instruction': parser.get_format_instructions()}
)

In [59]:
chain= template2 | groq_model | parser

chain.invoke({"topic": topic})

{'facts': [{'id': 1,
   'description': 'Intelligent agents can range from simple to highly complex, including systems like thermostats, control systems, and even human beings.'},
  {'id': 2,
   'description': 'Intelligent agents operate based on an objective function that encapsulates their goals and create plans to maximize the expected value of this function.'},
  {'id': 3,
   'description': 'Intelligent agents can be guided by various functions, such as reward functions in reinforcement learning or fitness functions in evolutionary algorithms.'},
  {'id': 4,
   'description': 'The concept of intelligent agents is studied in multiple fields, including artificial intelligence, economics, cognitive science, ethics, and philosophy of practical reason.'},
  {'id': 5,
   'description': 'Intelligent agents can be described as abstract functional systems, similar to computer programs, and are closely related to software agents and the concept of rational agents in economics.'}]}

## **Generate a Strcuture output**

In [60]:
from langchain.output_parsers import StructuredOutputParser, ResponseSchema

In [67]:
schema=[
    ResponseSchema(name="company_name", description="Name of compamy providing job in Job description"),
    ResponseSchema(name="Position", description="Position open in job deacription"),
    ResponseSchema(name="salary", description="Salary mentioned"),
    ResponseSchema(name="qualification", description="Qualification mentioned in job description"),
    ResponseSchema(name="Experience", description="Experience mentioned in job description"),
    ResponseSchema(name="Other Information", description="Other information mentioned in job description"),
]

parser=StructuredOutputParser.from_response_schemas(schema)

In [68]:
job_description = """
Lead India logo
Lead India
Share
Show more options
Back End Developer Trainee
India · 6 hours ago · Over 100 people clicked apply
Promoted by hirer · Responses managed off LinkedIn


 RemoteMatches your job preferences, workplace type is Remote.
Full-time
0 of 8 skills matchRequired skills are missing from your profile

Apply

Save
Save Back End Developer Trainee at Lead India
Back End Developer Trainee
Lead India · India (Remote)

Apply

Save
Save Back End Developer Trainee at Lead India
Show more options
How your profile and resume fit this job
Get AI-powered advice on this job and more exclusive features with Premium. Try Premium for ₹0



Tailor my resume to this job

Am I a good fit for this job?

How can I best position myself for this job?

About the job
Job Title: Back End Developer Trainee
Location: Remote
Job Type: Internship (Full-Time)
Duration: 1–3 Months
Stipend: ₹25,000/month
Department: Software Development / Engineering
Job Summary:

We are looking for a highly motivated and technically driven Back End Developer Trainee to join our remote team. This internship is ideal for individuals with a strong foundation in server-side development who are eager to gain hands-on experience working on real-world applications, APIs, and databases.
Key Responsibilities:

Assist in building and maintaining server-side logic, APIs, and databases
Write clean, scalable, and efficient code using languages such as Python, Node.js, or Java
Help integrate front-end elements with server-side logic
Work with databases (e.g., MySQL, PostgreSQL, MongoDB) to store and manage data
Participate in testing, debugging, and performance tuning of backend systems
Document code and backend processes for clarity and maintainability
Qualifications:

Bachelor’s degree (or final year student) in Computer Science, IT, or a related field
Solid understanding of one or more backend languages (Python, Node.js, Java, etc.)
Basic knowledge of RESTful API design and development
Familiarity with relational and/or NoSQL databases
Good problem-solving and analytical skills
Ability to work independently in a remote environment
Preferred Skills (Nice to Have):

Experience with version control systems like Git
Familiarity with frameworks (e.g., Express.js, Django, Spring Boot)
Understanding of authentication, authorization, and secure coding practices
Exposure to Docker, cloud platforms, or CI/CD pipelines is a plus
What We Offer:

Monthly stipend of ₹25,000
Remote work opportunity
Mentorship and training from experienced backend engineers
Involvement in real-world projects and live systems
Certificate of Completion
Potential for a full-time opportunity based on performance
"""

In [69]:
template3 = PromptTemplate(
    template='Give info from {job_description} \n {format_instruction}',
    input_variables=['job_description'],
    partial_variables={'format_instruction':parser.get_format_instructions()}
)

In [70]:
chain = template3 | groq_model | parser
chain.invoke({"job_description": job_description})

{'company_name': 'Lead India',
 'Position': 'Back End Developer Trainee',
 'salary': '₹25,000/month',
 'qualification': 'Bachelor’s degree (or final year student) in Computer Science, IT, or a related field',
 'Experience': 'No experience mentioned, but preferable skills include experience with version control systems like Git, familiarity with frameworks, and understanding of authentication, authorization, and secure coding practices',
 'Other Information': 'Remote work opportunity, 1-3 months duration, full-time internship, mentorship and training from experienced backend engineers, involvement in real-world projects and live systems, certificate of completion, and potential for a full-time opportunity based on performance'}

## **Pydantic Output Parse**

In [72]:
from pydantic import BaseModel, Field
from langchain_core.output_parsers import PydanticOutputParser

In [73]:
class Job_description(BaseModel):
    company_name:str=Field(description="name of the company")
    salary:int=Field(gt=18000,description="Salary in Job Description")
    remote:bool=Field(description="remote or not")

In [74]:
parser=PydanticOutputParser(pydantic_object=Job_description)

print(parser.get_format_instructions())

The output should be formatted as a JSON instance that conforms to the JSON schema below.

As an example, for the schema {"properties": {"foo": {"title": "Foo", "description": "a list of strings", "type": "array", "items": {"type": "string"}}}, "required": ["foo"]}
the object {"foo": ["bar", "baz"]} is a well-formatted instance of the schema. The object {"properties": {"foo": ["bar", "baz"]}} is not well-formatted.

Here is the output schema:
```
{"properties": {"company_name": {"description": "name of the company", "title": "Company Name", "type": "string"}, "salary": {"description": "Salary in Job Description", "exclusiveMinimum": 18000, "title": "Salary", "type": "integer"}, "remote": {"description": "remote or not", "title": "Remote", "type": "boolean"}}, "required": ["company_name", "salary", "remote"]}
```


In [75]:
template = PromptTemplate(
    template='Generate the infomation {job_description} person \n {format_instruction}',
    input_variables=['place'],
    partial_variables={'format_instruction':parser.get_format_instructions()}
)

In [79]:
chain = template | groq_model | parser

result=chain.invoke({"job_description": job_description})
print(result)

company_name='Lead India' salary=25000 remote=True


In [82]:
print(result.company_name)
print(result.salary)
print(result.remote)

Lead India
25000
True
